pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/assignment/segment/strategy/SegmentAssignmentStrategyFactory.java

```diff
@@ -0,0 +1,105 @@
+import java.util.Map;
+import org.apache.helix.HelixManager;
+import org.apache.pinot.common.assignment.InstancePartitions;
+import org.apache.pinot.spi.config.table.TableConfig;
+import org.apache.pinot.spi.config.table.TableType;
+import org.apache.pinot.spi.config.table.assignment.InstancePartitionsType;
+import org.apache.pinot.spi.config.table.assignment.SegmentAssignmentConfig;
+import org.apache.pinot.spi.utils.CommonConstants.Segment.AssignmentStrategy;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * Factory for SegmentAssignmentStrategy
+ */
+public class SegmentAssignmentStrategyFactory {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(SegmentAssignmentStrategyFactory.class);
+  private SegmentAssignmentStrategyFactory() {
+  }
+
+  /**
+   * Determine Segment Assignment strategy
+   */
+  public static SegmentAssignmentStrategy
+  getSegmentAssignmentStrategy(HelixManager helixManager, TableConfig tableConfig,
+      InstancePartitions instancePartitions, InstancePartitionsType instancePartitionsType) {
+    String assignmentStrategy = null;
+
+    TableType currentTableType = tableConfig.getTableType();
+    // TODO: Handle segment assignment strategy in future for CONSUMING segments in follow up PR
+    // See https://github.com/apache/pinot/issues/9047
+    // Accommodate new changes for assignment strategy
+    Map<String, SegmentAssignmentConfig>
+        segmentAssignmentConfigMap = tableConfig.getSegmentAssignmentConfigMap();
+
+    if (tableConfig.isDimTable()) {
+      // Segment Assignment Strategy for DIM tables
+      Preconditions.checkState(currentTableType == TableType.OFFLINE,
+          "DIM table Segment assignment Strategy is only applicable to OfflineTables");
+      SegmentAssignmentStrategy segmentAssignmentStrategy = new DimTableSegmentAssignmentStrategy();
+      segmentAssignmentStrategy.init(helixManager, tableConfig);
+      return segmentAssignmentStrategy;
+    } else {
+      // Try to determine segment assignment strategy from table config
+      if (segmentAssignmentConfigMap != null) {
+        SegmentAssignmentConfig segmentAssignmentConfig;
+        // Use the pre defined segment assignment strategy
+        segmentAssignmentConfig = segmentAssignmentConfigMap.get(instancePartitionsType.toString().toUpperCase());
+        // Segment assignment config is only applicable to offline tables and completed segments of real time tables
+        if (segmentAssignmentConfig != null) {
+          assignmentStrategy = segmentAssignmentConfig.getAssignmentStrategy();
+        }
+      }
+    }
+
+    // Use the existing information to determine segment assignment strategy
+    SegmentAssignmentStrategy segmentAssignmentStrategy;
+    if (assignmentStrategy == null) {
+      // Calculate numReplicaGroups and numPartitions to determine segment assignment strategy
+      int numReplicaGroups = instancePartitions.getNumReplicaGroups();
```

Comment:
I think that we will get the `null pointer error` if `instancePartitions = null`.

Reply:
We check null for instance partitions both in Realtime and offline segment assignment. I also added an extra preconditions for instance partitions to be not null in assignment strategy factory